{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run -i ../src/utils.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = PDBJsonParser('/home/sc.uni-leipzig.de/nq194gori/github/athaliana-struct/rcsb_pdb_custom_report_20230624064914.json')\n",
    "result = parser.parse_json()\n",
    "parser.save_to_markdown(\"summary.md\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser.get_repeated_database_accessions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = PDBJsonParser('/home/sc.uni-leipzig.de/nq194gori/github/athaliana-struct/rcsb_pdb_custom_report_20230624064914.json')\n",
    "unique_accessions = parser.get_unique_database_accessions()\n",
    "\n",
    "unique_sequences = {}\n",
    "for accession, data in unique_accessions.items():\n",
    "    # Save a dictionary with the accession as key and the sequence as value\n",
    "    unique_sequences[accession] = {data['sequence'] , data['pdb_id']}\n",
    "    #print(f\"Accession: {accession}, PDB ID: {data['pdb_id']}, Sequence: {data['sequence']}, \")\n",
    "\n",
    "unique_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required classes\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# Create an instance of PDBJsonParser and parse the JSON file\n",
    "parser = PDBJsonParser('/home/sc.uni-leipzig.de/nq194gori/github/athaliana-struct/rcsb_pdb_custom_report_20230624064914.json')\n",
    "\n",
    "# Get unique database accessions from the parsed JSON\n",
    "unique_accessions = parser.get_unique_database_accessions()\n",
    "\n",
    "unique_sequences = {}\n",
    "for accession, sequence in unique_accessions.items():\n",
    "    # Save a dictionary with the accession as key and the sequence as value\n",
    "    unique_sequences[accession] = sequence\n",
    "\n",
    "# Create an instance of AF2ProteinDownloader\n",
    "downloader = AF2ProteinDownloader()\n",
    "\n",
    "for accession, data in unique_sequences.items():\n",
    "    try:\n",
    "        print(f\"Downloading RCSB pdb for {data['pdb_id']}\")\n",
    "        # Download the pdb file from RCSB\n",
    "        downloader.download_rcsb_pdb(data['pdb_id'])\n",
    "\n",
    "        print(f\"Downloading AF2 pdb for {accession}\")\n",
    "        # Download the pdb file from AF2\n",
    "        downloader.download_pdb(accession)\n",
    "\n",
    "        print(f\"Downloading fasta for {accession}\")\n",
    "        # Download the fasta file\n",
    "        downloader.download_fasta(accession)\n",
    "\n",
    "        print(f\"Download complete for {accession}\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"ERROR: Could not download files for {accession}. Error: {str(e)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required library\n",
    "import os\n",
    "%run -i ../src/utils.py\n",
    "import sys  \n",
    "sys.path.insert(0, '../src/')\n",
    "import utils\n",
    "\n",
    "# Import required classes\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# Create an instance of PDBJsonParser and parse the JSON file\n",
    "parser = PDBJsonParser('../rcsb_pdb_custom_report_20230624064914.json')\n",
    "\n",
    "# Get unique database accessions from the parsed JSON\n",
    "unique_accessions = parser.get_unique_database_accessions()\n",
    "\n",
    "unique_sequences = {}\n",
    "for accession, sequence in unique_accessions.items():\n",
    "    # Save a dictionary with the accession as key and the sequence as value\n",
    "    unique_sequences[accession] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Import required library\n",
    "import os\n",
    "%run -i ../src/utils.py\n",
    "import sys  \n",
    "sys.path.insert(0, '../src/')\n",
    "import utils\n",
    "\n",
    "# Import required classes\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# Create an instance of PDBJsonParser and parse the JSON file\n",
    "parser = PDBJsonParser('../rcsb_pdb_custom_report_20230624064914.json')\n",
    "\n",
    "# Get unique database accessions from the parsed JSON\n",
    "unique_accessions = parser.get_unique_database_accessions()\n",
    "\n",
    "unique_sequences = {}\n",
    "for accession, sequence in unique_accessions.items():\n",
    "    # Save a dictionary with the accession as key and the sequence as value\n",
    "    unique_sequences[accession] = sequence\n",
    "\n",
    "\n",
    "for key, value in unique_sequences.items():\n",
    "    with open('all.fasta',\"w\") as file:\n",
    "        for key, value in unique_sequences.items():\n",
    "            file.write(f'>{key}\\n')\n",
    "            file.write(f'{value[\"sequence\"]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein Q96524 Sequence: MKMDKKTIVWFRRDLRIEDNPALAAAAHEGSVFPVFIWCPEEEGQFYPGRASRWWMKQSLAHLSQSLKALGSDLTLIQTHNTISAILDCIRVTGPTKVVFNHLYDPVSLVRDHTVKEKLVERGISVQSYNGDLLYEPWEIYCEKGKPFTSFNSYWKKCLDMSIESVMLPPPWRLMPITAAAEAIWACSIEELGLENEAEKPSNALLTRAWSPGWSNADKLLNEFIEKQLIDYAKNSKKVVGNSTSLLSPYLHFGEISVRHVFQCARMKQIIWARDKNSEGEESADLFLRGIGLREYSRYICFNFPFTHEQSLLSHLRFFPWDADVDKFKAWRQGRTGYPLVDAGMRELWATGWMHNRIRVIVSSFGVKFLLLPWKWGMKYFWDTLLDADLECDILGWQYISGSIPDGHELDRLDNPALQGAKYDPEGEYIRQWLPELARLPTEWIHHPWDAPLTVLKASGVELGTNYAKPIVDIDTARELLAKAISRTREAQIMIGAAPDEIVADSFEALGANTIKEPGLCPSVSSNDQQVPSVVRYNGSKRVKPEEEEERDMKKSRGFDERELFSTAESSSSSSVFFVSQSCSLASEGKNLEGIQDSSDQITTSLGKNGCK\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Import required library\n",
    "import os\n",
    "%run -i ../src/utils.py\n",
    "import sys  \n",
    "sys.path.insert(0, '../src/')\n",
    "import utils\n",
    "\n",
    "# Import required classes\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# Create an instance of PDBJsonParser and parse the JSON file\n",
    "parser = PDBJsonParser('../rcsb_pdb_custom_report_20230624064914.json')\n",
    "\n",
    "# Get unique database accessions from the parsed JSON\n",
    "unique_accessions = parser.get_unique_database_accessions()\n",
    "\n",
    "unique_sequences = {}\n",
    "for accession, sequence in unique_accessions.items():\n",
    "    # Save a dictionary with the accession as key and the sequence as value\n",
    "    unique_sequences[accession] = sequence\n",
    "\n",
    "import concurrent.futures\n",
    "\n",
    "# Set the directory where the protein data will be stored\n",
    "data_dir = \"proteins_data\"\n",
    "\n",
    "def process_protein(uniprot_id, data):\n",
    "    # Create an instance of the Protein class for each unique sequence\n",
    "    protein = Protein(uniprot_id=uniprot_id, \n",
    "                      pdb_id=data['pdb_id'],\n",
    "                      data_dir=data_dir, \n",
    "                      sequence_rcsb=data['sequence'])\n",
    "    \n",
    "    # Print the protein sequence\n",
    "    sequence = protein.sequence_rcsb\n",
    "    print(f\"Protein {uniprot_id} Sequence: {sequence}\")\n",
    "\n",
    "    # Download the protein structure from AF2\n",
    "    pdb_structure_af2 = protein.download_pdb_structures(\"AF2\")\n",
    "\n",
    "    # Download the protein structure from RCSB\n",
    "    pdb_structure_rcsb = protein.download_pdb_structures(\"RCSB\")\n",
    "\n",
    "    # Download the af2 fasta file\n",
    "    test = protein.sequence_uniprot()\n",
    "    \n",
    "    # Predict the structure of the protein using ESMFold\n",
    "    # Create FASTA file path\n",
    "    fasta_file = os.path.join(protein.protein_dir, f\"{protein.uniprot_id}.fasta\")\n",
    "\n",
    "    # Create output PDB file path for ESM predicted structure\n",
    "    esm_pdb_file = os.path.join(protein.protein_dir, f\"{protein.uniprot_id}_ESM\")\n",
    "\n",
    "    # Get predicted structure using ESM\n",
    "    protein.predict_structure_with_esm(fasta_file, esm_pdb_file)\n",
    "\n",
    "    # Now you should have the ESM predicted structure saved in the path defined by `esm_pdb_file`\n",
    "    print(f\"ESM predicted structure saved at: {esm_pdb_file}\")\n",
    "\n",
    "# Process Q96524 protein\n",
    "\n",
    "process_protein(\"Q96524\", unique_sequences[\"Q96524\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_sequences.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required library\n",
    "import os\n",
    "%run -i ../src/utils.py\n",
    "import sys  \n",
    "sys.path.insert(0, '../src/')\n",
    "import utils\n",
    "\n",
    "\n",
    "# Set the directory where the protein data will be stored\n",
    "data_dir = \"proteins_data\"\n",
    "\n",
    "# Create an instance of the Protein class using the UniProt ID of the protein EGFR\n",
    "protein = Protein(uniprot_id=\"Q8VZU4\", \n",
    "                pdb_id= unique_sequences['Q8VZU4']['pdb_id'],\n",
    "                data_dir=data_dir, \n",
    "                sequence_rcsb=unique_sequences['Q8VZU4']['sequence'])\n",
    "\n",
    "# Download the protein sequence from UniProt\n",
    "sequence = protein.sequence_rcsb\n",
    "print(\"Protein Sequence:\")\n",
    "print(sequence)\n",
    "\n",
    "# Download the protein structure from AF2\n",
    "pdb_structure_af2 = protein.download_pdb_structures(\"AF2\")\n",
    "\n",
    "# Download the protein structure from RCSB\n",
    "pdb_structure_rcsb = protein.download_pdb_structures(\"RCSB\")\n",
    "\n",
    "\n",
    "\n",
    "protein.calculate_rmsd_chimeraX()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = PDBJsonParser('/home/sc.uni-leipzig.de/nq194gori/github/athaliana-struct/rcsb_pdb_custom_report_20230624064914.json')\n",
    "unique_accessions = parser.get_unique_database_accessions()\n",
    "\n",
    "unique_sequences = {}\n",
    "for accession, data in unique_accessions.items():\n",
    "    # Save a dictionary with the accession as key and the sequence as value\n",
    "    unique_sequences[accession] = {data['sequence'] , data['pdb_id']}\n",
    "    #print(f\"Accession: {accession}, PDB ID: {data['pdb_id']}, Sequence: {data['sequence']}, \")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install bs4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the directory where the protein data will be stored\n",
    "data_dir = \"proteins_data\"\n",
    "\n",
    "# Iterate over the unique_sequences dictionary\n",
    "for uniprot_id, data in unique_sequences.items():\n",
    "    # Create an instance of the Protein class for each unique sequence\n",
    "    protein = Protein(uniprot_id=uniprot_id, \n",
    "                      pdb_id=data['pdb_id'],\n",
    "                      data_dir=data_dir, \n",
    "                      sequence_rcsb=data['sequence'])\n",
    "    \n",
    "    # Print the protein sequence\n",
    "    sequence = protein.sequence_rcsb\n",
    "    print(f\"Protein {uniprot_id} Sequence: {sequence}\")\n",
    "\n",
    "    # Download the protein structure from AF2\n",
    "    pdb_structure_af2 = protein.download_pdb_structures(\"AF2\")\n",
    "\n",
    "    # Download the protein structure from RCSB\n",
    "    pdb_structure_rcsb = protein.download_pdb_structures(\"RCSB\")\n",
    "    \n",
    "    # Calculate RMSD using ChimeraX\n",
    "    rmsd = protein.calculate_rmsd_chimeraX()\n",
    "    print(f\"RMSD for Protein {uniprot_id}: {rmsd}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count unique_sequences \n",
    "len(set(unique_sequences))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run in parallel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required library\n",
    "import os\n",
    "%run -i ../src/utils.py\n",
    "import sys  \n",
    "sys.path.insert(0, '../src/')\n",
    "import utils\n",
    "\n",
    "# Import required classes\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# Create an instance of PDBJsonParser and parse the JSON file\n",
    "parser = PDBJsonParser('../rcsb_pdb_custom_report_20230624064914.json')\n",
    "\n",
    "# Get unique database accessions from the parsed JSON\n",
    "unique_accessions = parser.get_unique_database_accessions()\n",
    "\n",
    "unique_sequences = {}\n",
    "for accession, sequence in unique_accessions.items():\n",
    "    # Save a dictionary with the accession as key and the sequence as value\n",
    "    unique_sequences[accession] = sequence\n",
    "\n",
    "import concurrent.futures\n",
    "\n",
    "# Set the directory where the protein data will be stored\n",
    "data_dir = \"proteins_data\"\n",
    "\n",
    "def process_protein(uniprot_id, data):\n",
    "    # Create an instance of the Protein class for each unique sequence\n",
    "    protein = Protein(uniprot_id=uniprot_id, \n",
    "                      pdb_id=data['pdb_id'],\n",
    "                      data_dir=data_dir, \n",
    "                      sequence_rcsb=data['sequence'])\n",
    "    \n",
    "    # Print the protein sequence\n",
    "    sequence = protein.sequence_rcsb\n",
    "    print(f\"Protein {uniprot_id} Sequence: {sequence}\")\n",
    "\n",
    "    # Download the protein structure from AF2\n",
    "    pdb_structure_af2 = protein.download_pdb_structures(\"AF2\")\n",
    "\n",
    "    # Download the protein structure from RCSB\n",
    "    pdb_structure_rcsb = protein.download_pdb_structures(\"RCSB\")\n",
    "    \n",
    "    # Predict the structure of the protein using ESMFold\n",
    "    # Create FASTA file path\n",
    "    fasta_file = os.path.join(protein.protein_dir, f\"{protein.uniprot_id}.fasta\")\n",
    "\n",
    "    # Create output PDB file path for ESM predicted structure\n",
    "    esm_pdb_file = os.path.join(protein.protein_dir, f\"{protein.uniprot_id}_ESM.pdb\")\n",
    "\n",
    "    # Get predicted structure using ESM\n",
    "    protein.predict_structure_with_esm(fasta_file, esm_pdb_file)\n",
    "\n",
    "    # Now you should have the ESM predicted structure saved in the path defined by `esm_pdb_file`\n",
    "    print(f\"ESM predicted structure saved at: {esm_pdb_file}\")\n",
    "\n",
    "    # Calculate rmsd using ChimeraX\n",
    "    #rmsd = protein.calculate_rmsd_chimeraX()\n",
    "    #print(f\"RMSD for Protein {uniprot_id}: {rmsd}\")\n",
    "    \n",
    "    return rmsd  # You can return whatever is necessary\n",
    "\n",
    "\n",
    "# Initialize a dictionary of dictionaries to store the returned results\n",
    "results = {}\n",
    "\n",
    "with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:\n",
    "    futures = {executor.submit(process_protein, uniprot_id, data): uniprot_id for uniprot_id, data in unique_sequences.items()}\n",
    "\n",
    "    for future in concurrent.futures.as_completed(futures):\n",
    "        uniprot_id = futures[future]\n",
    "        try:\n",
    "            result = future.result()\n",
    "            # Wait 1 second before processing the next protein\n",
    "            import time\n",
    "            time.sleep(1)\n",
    "            # Save the returned dictionary into the results dictionary\n",
    "            results[uniprot_id] = result\n",
    "        except Exception as exc:\n",
    "            print(f'{uniprot_id} generated an exception: {exc}')\n",
    "            # Save the exception to the error_log.txt file\n",
    "            with open('error_log.txt', 'a') as f:\n",
    "                f.write(f'{uniprot_id} generated an exception: {exc}\\n')\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# Create a pandas dataframe from the results dictionary\n",
    "results_df = pd.DataFrame(results).T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a summary of the results_df\n",
    "\n",
    "# Print the name of the columsn and their types\n",
    "\n",
    "#<class 'pandas.core.frame.DataFrame'>\n",
    "#Index: 574 entries, P93311 to Q9LK22\n",
    "#Data columns (total 12 columns):\n",
    "# #   Column                                   Non-Null Count  Dtype \n",
    "#---  ------                                   --------------  ----- \n",
    "# 0   Chain 1                                  574 non-null    object\n",
    "# 1   Chain 2                                  574 non-null    object\n",
    "# 2   Sequence alignment score                 574 non-null    object\n",
    "# 3   Alignment identifier                     574 non-null    object\n",
    "# 4   Chain 1 for RMSD                         574 non-null    object\n",
    "# 5   Chain 2 for RMSD                         574 non-null    object\n",
    "# 6   RMSD pruned                              574 non-null    object\n",
    "# 7   Number of CA atoms used for RMSD pruned  574 non-null    object\n",
    "# 8   RMSD all pairs                           574 non-null    object\n",
    "# 9   Number of CA atoms used for RMSD all     574 non-null    object\n",
    "# 10  Sequence 1                               574 non-null    object\n",
    "# 11  Sequence 2                               574 non-null    object\n",
    "#dtypes: object(12)\n",
    "#memory usage: 74.5+ KB\n",
    "import plotly.graph_objects as go\n",
    "# Sort numerically the CA columns\n",
    "results_df[\"Number of CA atoms used for RMSD pruned\"] = results_df[\"Number of CA atoms used for RMSD pruned\"].astype(int)\n",
    "results_df[\"Number of CA atoms used for RMSD all\"] = results_df[\"Number of CA atoms used for RMSD all\"].astype(int)\n",
    "# Save the results_df as a csv file\n",
    "results_df.to_csv(\"results.csv\")\n",
    "# Sort the dataframe by the CA columns\n",
    "#results_df = results_df.sort_values(by=[\"Number of CA atoms used for RMSD pruned\", \"Number of CA atoms used for RMSD all\"])\n",
    "from plotly.subplots import make_subplots\n",
    "fig = make_subplots(rows=2, cols=2, subplot_titles=('RMSD pruned', 'RMSD all pairs', 'CA atoms used for RMSD pruned', 'CA atoms used for RMSD all'))\n",
    "fig.add_trace(hist_rmsd_pruned, row=1, col=1)\n",
    "fig.add_trace(hist_rmsd_all_pairs, row=1, col=2)\n",
    "fig.add_trace(hist_ca_rmsd_pruned, row=2, col=1)\n",
    "fig.add_trace(hist_ca_rmsd_all, row=2, col=2)\n",
    "\n",
    "fig.update_layout(\n",
    "    height=800, \n",
    "    width=1400, \n",
    "    title_text=\"RMSD RCSB vs AlphaFold2DB for A. thaliana\",\n",
    "    \n",
    "    title_font=dict(size=24),\n",
    "    font=dict(size=18),\n",
    "    template='plotly_white',\n",
    "    showlegend=True,\n",
    "    # Define labels for y axis on left side (col=1)\n",
    "    yaxis_title=\"Frequency\",\n",
    "    # Define labels for y axis on right side (col=2)\n",
    "    yaxis2_title=\"Frequency\",\n",
    "    # Define labels for y axis on left side (col=1)\n",
    "    yaxis3_title=\"Frequency\",\n",
    "    # Define labels for y axis on right side (col=2)\n",
    "    yaxis4_title=\"Frequency\",\n",
    "    # Define labels for x axis on top side (row=1)\n",
    "    xaxis_title=\"RMSD bins\",\n",
    "    xaxis2_title=\"RMSD bins\",\n",
    "    # Define labels for x axis on bottom side (row=2)\n",
    "    xaxis3_title=\"Number of CA Bins\",\n",
    "    xaxis4_title=\"Number of CA Bins\"\n",
    ")\n",
    "\n",
    "# Update xaxis and yaxis parameters to black grid lines and plot border\n",
    "for i in range(1, 5):\n",
    "    fig.update_yaxes(\n",
    "        showgrid=True, \n",
    "        gridcolor='black', \n",
    "        showline=True, \n",
    "        linecolor='black',\n",
    "        showticklabels=True,  # Show tick labels on y-axis\n",
    "        tickfont=dict(color='black'),  # Set tick label color to black\n",
    "        title_standoff=10,  # Add more space between label and axis\n",
    "        row=i//3+1, \n",
    "        col=i%2+1\n",
    "    )\n",
    "    fig.update_xaxes(\n",
    "        showgrid=True, \n",
    "        gridcolor='black', \n",
    "        showline=True, \n",
    "        linecolor='black',\n",
    "        showticklabels=True,  # Show tick labels on x-axis\n",
    "        tickfont=dict(color='black'),  # Set tick label color to black\n",
    "        row=i//3+1, \n",
    "        col=i%2+1\n",
    "    )\n",
    "\n",
    "# Add black frame around each subplot\n",
    "fig.update_layout(\n",
    "    plot_bgcolor='white',  # Set plot background color to white\n",
    "    showlegend=True,\n",
    "    yaxis=dict(\n",
    "        showline=True,\n",
    "        linecolor='black',  # Set y-axis line color to black\n",
    "        ticks='outside',\n",
    "        tickcolor='black',  # Set y-axis tick color to black\n",
    "        tickfont=dict(color='black'),  # Set y-axis tick label color to black\n",
    "        title_standoff=15,  # Add more space between label and axis\n",
    "        title_font=dict(color='black')  # Set y-axis label color to black\n",
    "    ),\n",
    ")\n",
    "# Center title\n",
    "fig.update_layout(title_x=0.5)\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n",
    "\n",
    "# Save the plot as html file\n",
    "fig.write_html(\"rmsd.html\")\n",
    "\n",
    "# Save as a pdf file\n",
    "fig.write_image(\"rmsd.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
